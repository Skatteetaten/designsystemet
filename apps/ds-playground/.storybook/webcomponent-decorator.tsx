import { Decorator } from '@storybook/react';
import { createRoot, Root } from 'react-dom/client';

export interface WebcompAttributes {
  [key: string]: string | undefined;
}
export const webComponent: Decorator = function (Story, context) {
  const Name = context.parameters['customElementName'] ?? 'ds-customelement';
  class WebComponent extends HTMLElement {
    mountPoint!: HTMLElement;
    props!: WebcompAttributes; // Frivillige attributes som custom-elementet kan ta imot - de oversettes til props
    root!: Root;

    connectedCallback(): void {
      this.mountReactComponent();
    }

    disconnectedCallback(): void {
      // setTimeout for å gjøre unmount async. Ellers får vi en feilmelding
      // Warning: Attempted to synchronously unmount a root while React was already rendering.
      // React cannot finish unmounting the root until the current render has completed,
      // which may lead to a race condition.
      setTimeout(() => {
        this.root.unmount();
      });
    }

    // henter custom-elementets attributer - de assigneres til props lengre ned
    get value(): WebcompAttributes {
      return Array.from(this.attributes).reduce(
        (previousValue: { [key: string]: string }, currentValue) => {
          previousValue[currentValue.name] = currentValue.value;
          return previousValue;
        },
        {}
      ) as WebcompAttributes;
    }

    // håndtering av events - disse skal sendes videre til komponentene gjennom bruk av EventForwarder wrappper
    eventDispatcher = (event: Event): void => {
      this.dispatchEvent(event);
    };

    subscribeDispatcher = (
      eventname: string,
      callback: (event: Event) => void
    ): void => {
      this.addEventListener(eventname, (e) => {
        return callback(e);
      });
    };

    mountReactComponent(): void | null {
      this.props = this.value;
      this.mountPoint = document.createElement('div');

      const shadowRoot = this.attachShadow({ mode: 'open' });

      shadowRoot.appendChild(this.mountPoint);

      if (import.meta.env.MODE === 'development') {
        for (const tag of document.head.querySelectorAll('style')) {
          const clone = tag.cloneNode(true);
          shadowRoot.appendChild(clone);
        }
      } else {
        const cssLink = document.createElement('link');
        cssLink.href = 'main.css';
        cssLink.rel = 'stylesheet';
        shadowRoot.appendChild(cssLink);
      }

      this.root = createRoot(this.mountPoint);
      this.root.render(<Story {...this.props} />);
    }
  }

  if (!window.customElements.get(Name)) {
    window.customElements.define(Name, WebComponent);
  }

  return <Name />;
};
