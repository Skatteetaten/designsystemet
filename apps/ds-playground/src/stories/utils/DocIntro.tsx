import { ReactNode } from 'react';

import { Link } from '@skatteetaten/ds-buttons';
import { Paragraph } from '@skatteetaten/ds-typography';

interface DocIntroProps {
  name: string;
  url: string;
}

export const DocIntro = ({ name, url }: DocIntroProps): ReactNode => {
  return (
    <Paragraph variant={'ingress'}>
      {`Se `}
      <Link href={url}>{name}</Link>
      {` p√• stil og tone for eksempler og
        retningslinjer for denne komponenten.`}
    </Paragraph>
  );
};
