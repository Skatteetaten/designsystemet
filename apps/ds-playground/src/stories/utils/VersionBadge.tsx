import { ReactNode } from 'react';

interface VersionBadgeProps {
  packageName: string;
}

export const VersionBadge = ({ packageName }: VersionBadgeProps): ReactNode => {
  let color = 'green';

  let version =
    import.meta.env.DS_VERSIONS[`${packageName.replaceAll('-', '_')}`] ?? '..';
  if (import.meta.env.MODE === 'development') {
    color = 'red';
    version = 'developing';
  }
  if (import.meta.env['STORYBOOK_BUILD_VERSION']?.includes('SNAPSHOT')) {
    color = 'blue';
  }

  const normalizedPackageName = packageName.replaceAll('-', '--');
  const normalizedVersion = version.replaceAll('-', '--').replaceAll('_', '__');

  return (
    <img
      src={`https://img.shields.io/badge/%40skatteetaten%2Fds--${normalizedPackageName}-${normalizedVersion}-${color}?logo=npm`}
      alt={`@skatteetaten/ds-${packageName}@${version}`}
    />
  );
};
