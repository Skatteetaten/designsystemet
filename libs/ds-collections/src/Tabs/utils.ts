export const valueRegex = /^[A-Za-z][\w\-.:]*$/;
