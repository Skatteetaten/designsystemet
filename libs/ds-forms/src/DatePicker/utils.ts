import { format, isValid, parse } from 'date-fns';

export const formatDateForInput = (
  dateFormat: string,
  date?: Date | null
): string => {
  return !isValid(date) ? '' : format(date as Date, dateFormat);
};

const COMMON_INPUT_FORMATS = [
  'dd.MM.yy',
  'dd/MM/yy',
  'dd-MM-yy',
  'dd.MM.yyyy',
  'dd/MM/yyyy',
  'dd-MM-yyyy',
  'ddMM',
  'ddMMyy',
  'ddMMyyyy',
];

export const parseDateFromInput = (
  value: string,
  dateFormat: string
): Date | null => {
  const dateFormats = Array.from(
    new Set([...COMMON_INPUT_FORMATS, dateFormat])
  );

  for (const format of dateFormats) {
    const date = parse(value, format, new Date());
    if (isValid(date)) {
      return date;
    }
  }

  return null;
};
