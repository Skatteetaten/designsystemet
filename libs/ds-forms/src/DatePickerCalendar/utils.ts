import { dsI18n } from '@skatteetaten/ds-core-utils';
import {
  addDays,
  getWeekOfMonth,
  isAfter,
  isBefore,
  isSunday,
  isToday,
} from 'date-fns';

const lastValidYear = 9999;
const sunday = 0;

function getDaysFromPrevMonth(datesInSelectedMonth: Date[]): number {
  const firstDayInSelectedMonth = datesInSelectedMonth[0].getDay();
  let daysFromPrevMonthInFirstWeekOfSelectedMonth;
  if (firstDayInSelectedMonth === sunday) {
    daysFromPrevMonthInFirstWeekOfSelectedMonth = 6;
  } else {
    daysFromPrevMonthInFirstWeekOfSelectedMonth = firstDayInSelectedMonth - 1;
  }

  return daysFromPrevMonthInFirstWeekOfSelectedMonth;
}

function getDaysFromNextMonth(datesInSelectedMonth: Date[]): number {
  const lastDayInSelectedMonth =
    datesInSelectedMonth[datesInSelectedMonth.length - 1].getDay();
  let daysFromNextMonthInLastWeekOfSelectedMonth;
  if (lastDayInSelectedMonth === sunday) {
    daysFromNextMonthInLastWeekOfSelectedMonth = 0;
  } else {
    daysFromNextMonthInLastWeekOfSelectedMonth = 7 - lastDayInSelectedMonth;
  }

  return daysFromNextMonthInLastWeekOfSelectedMonth;
}

function getLastDateInPrevMonth(year: number, monthIndex: number): Date {
  const lastDateInPrevMonth = new Date(year, monthIndex, 0);
  lastDateInPrevMonth.setFullYear(lastDateInPrevMonth.getFullYear());

  return lastDateInPrevMonth;
}

function getFirstDateInNextMonth(year: number, monthIndex: number): Date {
  const firstDateInNextMonth = new Date(year, monthIndex + 1, 1);
  firstDateInNextMonth.setFullYear(firstDateInNextMonth.getFullYear());

  return firstDateInNextMonth;
}

interface Cell {
  date: Date;
  isAdjacentMonth: boolean;
  isToday: boolean;
  text: string;
  disabled: boolean;
}

function getCalendarCells(
  year: number,
  monthIndex: number,
  minDate?: Date,
  maxDate?: Date,
  disabledDates?: Date[]
): Cell[] {
  const dates = getDatesInMonth(year, monthIndex);
  const cells: Cell[] = [];

  const daysFromPrevMonth = getDaysFromPrevMonth(dates);
  const lastDateInPrevMonth = getLastDateInPrevMonth(year, monthIndex);

  for (let i = 0; i < daysFromPrevMonth; i++) {
    const date = new Date(lastDateInPrevMonth);

    const cell: Cell = {
      date,
      isAdjacentMonth: true,
      disabled:
        disabledDates?.some((d) => d.getTime() === date.getTime()) ||
        isNotInAllowedRange(date, minDate, maxDate),
      isToday: isToday(date),
      text: String(date.getDate()),
    };

    cells.unshift(cell);

    lastDateInPrevMonth.setDate(lastDateInPrevMonth.getDate() - 1);
  }

  dates.forEach((date) => {
    const cell: Cell = {
      date,
      isAdjacentMonth: false,
      disabled:
        disabledDates?.some((d) => d.getTime() === date.getTime()) ||
        isNotInAllowedRange(date, minDate, maxDate),
      isToday: isToday(date),
      text: String(date.getDate()),
    };

    cells.push(cell);
  });

  const daysFromNextMonth = getDaysFromNextMonth(dates);
  const firstDateInNextMonth = getFirstDateInNextMonth(year, monthIndex);
  for (let i = 0; i < daysFromNextMonth; i++) {
    const date = new Date(firstDateInNextMonth);

    const cell: Cell = {
      date,
      isAdjacentMonth: true,
      disabled:
        disabledDates?.some((d) => d.getTime() === date.getTime()) ||
        isNotInAllowedRange(date, minDate, maxDate),
      isToday: isToday(date),
      text: String(date.getDate()),
    };

    cells.push(cell);

    firstDateInNextMonth.setDate(firstDateInNextMonth.getDate() + 1);
  }

  return cells;
}

function getDatesInMonth(year: number, monthIndex: number): Date[] {
  const date = new Date(year, monthIndex);
  date.setFullYear(year);
  date.setDate(1);

  const month = date.getMonth();
  const dates: Date[] = [];
  while (date.getMonth() === month) {
    dates.push(new Date(date));
    date.setDate(date.getDate() + 1);
    date.setHours(0, 0, 0);
  }
  return dates;
}

export function getCalendarRows(
  year: number | string,
  monthIndex: number,
  minDate?: Date,
  maxDate?: Date,
  disabledDates?: Date[]
): Array<Cell>[] {
  year = findValidYear(year);
  const cells = getCalendarCells(
    year,
    monthIndex,
    minDate,
    maxDate,
    disabledDates
  );
  const rows: Array<Cell>[] = [];
  for (let i = 0; i < cells.length; i += 7) {
    rows.push(cells.slice(i, i + 7));
  }

  return rows;
}

export const isNotInAllowedRange = (
  date: Date,
  minDate?: Date,
  maxDate?: Date
): boolean => {
  minDate && minDate.setHours(0, 0, 0);
  return (
    (minDate ? isBefore(date, minDate) : false) ||
    (maxDate ? isAfter(date, maxDate) : false) ||
    date.getFullYear() > lastValidYear
  );
};

export const getNameOfMonthsAndDays = (): {
  monthNames: string[];
  dayNames: string[];
} => {
  return {
    monthNames: [
      dsI18n.t('ds_forms:datepicker.January'),
      dsI18n.t('ds_forms:datepicker.February'),
      dsI18n.t('ds_forms:datepicker.March'),
      dsI18n.t('ds_forms:datepicker.April'),
      dsI18n.t('ds_forms:datepicker.May'),
      dsI18n.t('ds_forms:datepicker.June'),
      dsI18n.t('ds_forms:datepicker.July'),
      dsI18n.t('ds_forms:datepicker.August'),
      dsI18n.t('ds_forms:datepicker.September'),
      dsI18n.t('ds_forms:datepicker.October'),
      dsI18n.t('ds_forms:datepicker.November'),
      dsI18n.t('ds_forms:datepicker.December'),
    ],
    dayNames: [
      dsI18n.t('ds_forms:datepicker.Mon'),
      dsI18n.t('ds_forms:datepicker.Tue'),
      dsI18n.t('ds_forms:datepicker.Wed'),
      dsI18n.t('ds_forms:datepicker.Thu'),
      dsI18n.t('ds_forms:datepicker.Fri'),
      dsI18n.t('ds_forms:datepicker.Sat'),
      dsI18n.t('ds_forms:datepicker.Sun'),
    ],
  };
};

export function findValidYear(year: string | number): number {
  return year === '' || year === 0 ? new Date().getFullYear() : Number(year);
}

export function getGridIdxForDate(focusableDate: Date): string {
  const colIdx = isSunday(focusableDate) ? 6 : focusableDate.getDay() - 1;
  const rowIdx = getWeekOfMonth(focusableDate, { weekStartsOn: 1 }) - 1;

  return `${rowIdx}${colIdx}`;
}

export const getFirstFocusableDate = (
  selectedDate: Date,
  minDate?: Date,
  maxDate?: Date
): Date => {
  if (isNotInAllowedRange(selectedDate, minDate, maxDate)) {
    let focusableDate = undefined;
    if (maxDate && isAfter(selectedDate, maxDate)) {
      focusableDate = maxDate;
    } else if (minDate && isBefore(selectedDate, minDate)) {
      focusableDate = minDate;
    }

    if (
      focusableDate &&
      !isNotInAllowedRange(focusableDate, minDate, maxDate)
    ) {
      return focusableDate;
    }
  }

  return selectedDate;
};

export const findNextAvailableDate = (
  startDate: Date,
  disabledDates?: Date[]
): Date => {
  if (!disabledDates || disabledDates.length === 0) {
    return addDays(startDate, 1);
  }

  const maxDate = new Date().setFullYear(lastValidYear);
  const disabledTimestamps = new Set(
    disabledDates.map((date) => date.getTime())
  );

  let currentDate = addDays(startDate, 1);
  let currentTimestamp = currentDate.getTime();

  while (
    disabledTimestamps.has(currentTimestamp) &&
    isBefore(currentDate, maxDate)
  ) {
    currentDate = addDays(currentDate, 1);
    currentTimestamp = currentDate.getTime();
  }

  return currentDate;
};

export const findPreviousAvailableDate = (
  startDate: Date,
  disabledDates?: Date[]
): Date => {
  if (!disabledDates || disabledDates.length === 0) {
    return addDays(startDate, -1);
  }

  const minDate = new Date('1000-01-01');
  const disabledTimestamps = new Set(
    disabledDates.map((date) => date.getTime())
  );

  let currentDate = addDays(startDate, -1);
  let currentTimestamp = currentDate.getTime();

  while (
    disabledTimestamps.has(currentTimestamp) &&
    isBefore(minDate, currentDate)
  ) {
    currentDate = addDays(currentDate, -1);
    currentTimestamp = currentDate.getTime();
  }

  return currentDate;
};
