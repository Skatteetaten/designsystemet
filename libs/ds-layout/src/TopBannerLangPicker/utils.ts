import { Languages } from '@skatteetaten/ds-core-utils';

import { TopBannerLangPickerProps } from './TopBannerLangPicker.types';

export const convertLocaleToLang = (locale: Languages): string => {
  return locale.substring(0, locale.indexOf('_'));
};

export type LanguageItem = {
  lang: string;
  displayName: string;
};

export type LanguageItems = {
  [key: string]: LanguageItem;
};

export const isLanguages = (value: unknown): value is Languages => {
  return Object.values(Languages).includes(value as Languages);
};

export const getCurrentLanguages = (
  includeSami: boolean,
  additionalLanguages: TopBannerLangPickerProps['additionalLanguages'] = []
): LanguageItems => {
  const defaultLanguages: LanguageItems = {
    nb: {
      lang: 'nb',
      displayName: 'Bokmål',
    },
    nn: {
      lang: 'nn',
      displayName: 'Nynorsk',
    },
    en: {
      lang: 'en',
      displayName: 'English',
    },
  };
  const se = {
    se: {
      lang: 'se',
      displayName: 'Sámegiella',
    },
  };

  const additionalLanguagesAsObject =
    additionalLanguages?.reduce<LanguageItems>((acc, curr) => {
      acc[curr.lang] = curr;
      return acc;
    }, {});

  return includeSami
    ? { ...defaultLanguages, ...se, ...additionalLanguagesAsObject }
    : { ...defaultLanguages, ...additionalLanguagesAsObject };
};
