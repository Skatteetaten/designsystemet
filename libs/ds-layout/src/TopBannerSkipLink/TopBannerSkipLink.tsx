import { forwardRef, SyntheticEvent, JSX } from 'react';
import { useTranslation } from 'react-i18next';

import { dsI18n, getCommonClassNameDefault } from '@skatteetaten/ds-core-utils';
import { ArrowDownSVGpath, Icon } from '@skatteetaten/ds-icons';

import { getTopBannerSkipLinkTargetDefault } from './defaults';
import { TopBannerSkipLinkProps } from './TopBannerSkipLink.types';

import styles from './TopBannerSkipLink.module.scss';

export const TopBannerSkipLink = forwardRef<
  HTMLAnchorElement,
  TopBannerSkipLinkProps
>(
  (
    {
      id,
      className = getCommonClassNameDefault(),
      lang,
      'data-testid': dataTestId,
      shadowRootNode,
      target = getTopBannerSkipLinkTargetDefault(),
      children,
    },
    ref
  ): JSX.Element => {
    const { t } = useTranslation('ds_layout', { i18n: dsI18n });

    const handleClick = (event: SyntheticEvent): void => {
      event.preventDefault();

      const container: HTMLElement | null = shadowRootNode
        ? shadowRootNode.querySelector(target)
        : document.querySelector(target);
      if (container) {
        container.tabIndex = -1;
        container.focus();
      }
    };

    return (
      <a
        ref={ref}
        id={id}
        href={target}
        className={`${styles.skipLink} ${className ?? ''}`.trim()}
        lang={lang}
        data-testid={dataTestId}
        data-skip-link={'ds-skip-link'}
        onClick={handleClick}
      >
        <span className={styles.skipLinkContent}>
          <Icon className={styles.icon} svgPath={ArrowDownSVGpath} />
          {children ?? t('topbanner.SkipLinkText')}
        </span>
      </a>
    );
  }
);
TopBannerSkipLink.displayName = 'TopBannerSkipLink';
