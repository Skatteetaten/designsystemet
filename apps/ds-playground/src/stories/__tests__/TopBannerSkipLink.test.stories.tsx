import { Meta, StoryFn, StoryObj } from '@storybook/react';
import { expect, fireEvent, userEvent, within } from '@storybook/test';

import { dsI18n } from '@skatteetaten/ds-core-utils';

import { wrapper } from './testUtils/storybook.testing.utils';
// eslint-disable-next-line @nx/enforce-module-boundaries
import { TopBannerSkipLink } from '../../../../../libs/ds-layout/src/TopBannerSkipLink/TopBannerSkipLink';
// eslint-disable-next-line @nx/enforce-module-boundaries
import { TopBannerSkipLinkProps } from '../../../../../libs/ds-layout/src/TopBannerSkipLink/TopBannerSkipLink.types';
import { webComponent } from '../../../.storybook/webcomponent-decorator';

const meta = {
  component: TopBannerSkipLink,
  title: 'Tester/TopBanner/TopBannerSkipLink (intern)',
  argTypes: {
    // Baseprops
    key: { table: { disable: true } },
    ref: { table: { disable: true } },
    className: { table: { disable: true } },
    id: { table: { disable: true } },
    lang: { table: { disable: true } },
    'data-testid': { table: { disable: true } },
    // Props
    children: { table: { disable: true } },
    shadowRootNode: {
      table: { disable: true },
    },
    target: { table: { disable: true } },
  },
} satisfies Meta<typeof TopBannerSkipLink>;
export default meta;
type Story = StoryObj<typeof meta>;

const skipLinkText = 'ds_layout:topbanner.SkipLinkText';
const defaultArgs: TopBannerSkipLinkProps = {};

export const WithRef = {
  name: 'With Ref (FA1)',
  args: {
    ref: (instance: HTMLAnchorElement | null): void => {
      if (instance) {
        instance.id = 'dummyIdForwardedFromRef';
      }
    },
  },
  argTypes: {
    ref: { table: { disable: false } },
  },
  parameters: {
    imageSnapshot: { disable: true },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const skipLink = canvas.getByRole('link');
    await expect(skipLink).toHaveAttribute('id', 'dummyIdForwardedFromRef');
  },
} satisfies Story;

export const WithAttributes = {
  name: 'With Attributes (FA2-5)',
  args: {
    id: 'htmlId',
    className: 'dummyClassname',
    lang: 'nb',
    'data-testid': '123ID',
  },
  argTypes: {
    id: { table: { disable: false } },
    className: { table: { disable: false } },
    lang: { table: { disable: false } },
    'data-testid': { table: { disable: false } },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const skipLink = canvas.getByRole('link');
    await expect(skipLink).toHaveAttribute('id', 'htmlId');
    await expect(skipLink).toHaveClass('dummyClassname');
    await expect(skipLink).toHaveAttribute('lang', 'nb');
    await expect(skipLink).toHaveAttribute('data-testid', '123ID');
    skipLink.focus();
  },
} satisfies Story;

export const Defaults = {
  name: 'Defaults (B2)',
  args: {
    ...defaultArgs,
  },
  argTypes: {},
  parameters: {
    imageSnapshot: {
      focus: `${wrapper} a`,
    },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const skipLink = canvas.getByRole('link', {
      name: dsI18n.t(skipLinkText),
    });
    expect(skipLink).toBeInTheDocument();
    expect(skipLink.tagName).toBe('A');
  },
} satisfies Story;

const mainId = 'hovedinnhold';
const MainTemplate: StoryFn<typeof TopBannerSkipLink> = (args) => (
  <>
    <TopBannerSkipLink {...args} />
    <main id={mainId}>{'Hovedinnholdet på siden'}</main>
  </>
);

export const WithTargetAndChildren = {
  render: MainTemplate,
  name: 'With Target And Children (A3 delvis)',
  args: {
    ...defaultArgs,
    target: `#${mainId}`,
    children: 'Snarvei til hovedinnhold',
  },
  argTypes: {
    target: { table: { disable: false } },
    children: { table: { disable: false } },
  },
  parameters: {
    imageSnapshot: { disable: true },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const skipLink = canvas.getByRole('link', {
      name: 'Snarvei til hovedinnhold',
    });
    await expect(skipLink).toBeInTheDocument();
    await expect(skipLink).toHaveAttribute('href', `#${mainId}`);
    await fireEvent.click(skipLink);
    const main = canvas.getByRole('main');
    await expect(main).toHaveFocus();
  },
} satisfies Story;

export const ClickSkipLink = {
  render: MainTemplate,
  name: 'Click SkipLink (A3, C1 delvis)',
  args: {
    ...defaultArgs,
  },
  argTypes: {},
  parameters: {
    imageSnapshot: { disable: true },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const main = canvas.getByRole('main');
    const skipLink = canvas.getByRole('link', {
      name: dsI18n.t(skipLinkText),
    });
    skipLink.focus();
    await fireEvent.click(skipLink);
    await expect(main).toHaveFocus();
  },
} satisfies Story;

const TemplateWithShadowDom: StoryFn<typeof TopBannerSkipLink> = (args) => {
  const element = document.querySelector('skiplink-customelement');
  const shadowRoot = element?.shadowRoot;
  return (
    <div>
      <TopBannerSkipLink {...args} shadowRootNode={shadowRoot ?? undefined} />
      <main>{'Hovedinnholdet på siden'}</main>
    </div>
  );
};

export const WithShadowDom = {
  render: TemplateWithShadowDom,
  name: 'With ShadowDom',
  args: {
    ...defaultArgs,
  },
  decorators: [webComponent],
  parameters: {
    imageSnapshot: {
      disable: true,
    },
    customElementName: 'skiplink-customelement',
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    //skipLink finnes ikke utenfor shadowDom
    await expect(canvas.queryByRole('link')).not.toBeInTheDocument();

    const customElement = canvasElement.querySelector(`skiplink-customelement`);
    await expect(customElement).toBeInTheDocument();
    const skipLink =
      customElement?.shadowRoot && customElement.shadowRoot.querySelector('a');

    await expect(skipLink).toBeInTheDocument();
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    await userEvent.click(skipLink!);

    const main =
      customElement?.shadowRoot &&
      customElement.shadowRoot.querySelector('main:focus');
    await expect(main).toBeInTheDocument();
  },
} satisfies Story;
