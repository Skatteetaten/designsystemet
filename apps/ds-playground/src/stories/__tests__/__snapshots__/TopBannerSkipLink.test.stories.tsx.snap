// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tester/TopBanner/TopBannerSkipLink (intern) ClickSkipLink play-test 1`] = `
<div data-test-block="true">
  <a href="main"
     class="_skipLink_nmfvk_1"
  >
    <span class="_skipLinkContent_nmfvk_22">
      <svg class="_systemIcon_dry7n_2 _systemIcon_medium_dry7n_10 _icon_nmfvk_45"
           viewbox="0 0 24 24"
           focusable="false"
           aria-hidden="true"
           role="img"
      >
        <path d="M4 14.64L12 21.5L20 14.64L18.33 12.7L13.25 17.05L13.25 3.5H10.75L10.75 17.05L5.67 12.7L4 14.64Z">
        </path>
      </svg>
      Hopp til hovedinnhold
    </span>
  </a>
  <main id="hovedinnhold"
        tabindex="-1"
  >
    Hovedinnholdet på siden
  </main>
</div>
`;

exports[`Tester/TopBanner/TopBannerSkipLink (intern) Defaults play-test 1`] = `
<div data-test-block="true">
  <a href="main"
     class="_skipLink_nmfvk_1"
  >
    <span class="_skipLinkContent_nmfvk_22">
      <svg class="_systemIcon_dry7n_2 _systemIcon_medium_dry7n_10 _icon_nmfvk_45"
           viewbox="0 0 24 24"
           focusable="false"
           aria-hidden="true"
           role="img"
      >
        <path d="M4 14.64L12 21.5L20 14.64L18.33 12.7L13.25 17.05L13.25 3.5H10.75L10.75 17.05L5.67 12.7L4 14.64Z">
        </path>
      </svg>
      Hopp til hovedinnhold
    </span>
  </a>
</div>
`;

exports[`Tester/TopBanner/TopBannerSkipLink (intern) WithAttributes play-test 1`] = `
<div data-test-block="true">
  <a id="htmlId"
     href="main"
     class="_skipLink_nmfvk_1 dummyClassname"
     lang="nb"
     data-testid="123ID"
  >
    <span class="_skipLinkContent_nmfvk_22">
      <svg class="_systemIcon_dry7n_2 _systemIcon_medium_dry7n_10 _icon_nmfvk_45"
           viewbox="0 0 24 24"
           focusable="false"
           aria-hidden="true"
           role="img"
      >
        <path d="M4 14.64L12 21.5L20 14.64L18.33 12.7L13.25 17.05L13.25 3.5H10.75L10.75 17.05L5.67 12.7L4 14.64Z">
        </path>
      </svg>
      Hopp til hovedinnhold
    </span>
  </a>
</div>
`;

exports[`Tester/TopBanner/TopBannerSkipLink (intern) WithRef play-test 1`] = `
<div data-test-block="true">
  <a href="main"
     class="_skipLink_nmfvk_1"
     id="dummyIdForwardedFromRef"
  >
    <span class="_skipLinkContent_nmfvk_22">
      <svg class="_systemIcon_dry7n_2 _systemIcon_medium_dry7n_10 _icon_nmfvk_45"
           viewbox="0 0 24 24"
           focusable="false"
           aria-hidden="true"
           role="img"
      >
        <path d="M4 14.64L12 21.5L20 14.64L18.33 12.7L13.25 17.05L13.25 3.5H10.75L10.75 17.05L5.67 12.7L4 14.64Z">
        </path>
      </svg>
      Hopp til hovedinnhold
    </span>
  </a>
</div>
`;

exports[`Tester/TopBanner/TopBannerSkipLink (intern) WithShadowDom play-test 1`] = `
<div data-test-block="true">
  <skiplink-customelement>
  </skiplink-customelement>
</div>
`;

exports[`Tester/TopBanner/TopBannerSkipLink (intern) WithTargetAndChildren play-test 1`] = `
<div data-test-block="true">
  <a href="#hovedinnhold"
     class="_skipLink_nmfvk_1"
  >
    <span class="_skipLinkContent_nmfvk_22">
      <svg class="_systemIcon_dry7n_2 _systemIcon_medium_dry7n_10 _icon_nmfvk_45"
           viewbox="0 0 24 24"
           focusable="false"
           aria-hidden="true"
           role="img"
      >
        <path d="M4 14.64L12 21.5L20 14.64L18.33 12.7L13.25 17.05L13.25 3.5H10.75L10.75 17.05L5.67 12.7L4 14.64Z">
        </path>
      </svg>
      Snarvei til hovedinnhold
    </span>
  </a>
  <main id="hovedinnhold"
        tabindex="-1"
  >
    Hovedinnholdet på siden
  </main>
</div>
`;
